### PR TITLE
Add manifest

### PR DESCRIPTION
**Pull request by GanZiheng**
Based on
- https://github.com/tikv/agatedb/pull/60
- https://github.com/dgraph-io/badger/blob/45bca18f24ef5cc04701a1e17448ddfce9372da0/manifest.go.

Signed-off-by: GanZiheng [ganziheng98@gmail.com](mailto:ganziheng98@gmail.com)

### CODE REVIEW

**Comment by skyzh** [comment]
We already have a complete implementation at develop branch 🤪

https://github.com/tikv/agatedb/blob/develop/src/manifest.rs

**Comment by BusyJay** [comment]
> We already have a complete implementation at develop branch 🤪
> 
> https://github.com/tikv/agatedb/blob/develop/src/manifest.rs

Indeed. @GanZiheng why not use that directly?

@skyzh do you have any suggestions on how to merge code from develop branch to master?

**Comment by skyzh** [comment]
At first glance, the code seems quite similar to develop branch. I’ll do a review and try recalling what I’ve written years ago. 😪😇

File: src/manifest.rs [hunk 1]
```diff
@@ -0,0 +1,442 @@
...
+    fn as_changes(&self) -> Vec<ManifestChange> {
+        let mut changes = Vec::with_capacity(self.tables.len());
+        for (id, tm) in &self.tables {
+            changes.push(new_create_change(*id, tm.level as usize, tm.key_id));
+        }
+        changes
+    }
+
+    pub fn replay(file: &mut File) -> Result<(Manifest, u32)> {
+        file.seek(SeekFrom::Start(0))?;
```
Comment (skyzh) [inline src/manifest.rs:95]: Better to wrap a BufReader with file when replaying.

File: src/manifest.rs [hunk 2]
```diff
@@ -0,0 +1,442 @@
...
+        }
+
+        let stat = file.metadata()?;
+        let mut build = Manifest::new();
+        let mut buf = vec![];
+        let mut len_crc_buf = vec![0; 8];
+        let mut offset = 8;
+
+        loop {
+            if file.read(&mut len_crc_buf)? != 8 {
```
Comment (skyzh) [inline src/manifest.rs:113]: I guess read function is not guaranteed to read 8 bytes. Maybe we should use `read_exact` here, and find EOF error to break.

File: src/manifest.rs [hunk 3]
```diff
@@ -0,0 +1,442 @@
...
+            core.rewrite(&self.directory)?;
+        } else {
+            let mut len_crc_buf = vec![0; 8];
+            (&mut len_crc_buf[..4]).put_u32(buf.len() as u32);
+            (&mut len_crc_buf[4..]).put_u32(crc32::checksum_castagnoli(&buf));
+            len_crc_buf.extend_from_slice(&buf);
+            write_with_length_check!(core.file.as_mut().unwrap(), len_crc_buf);
+        }
+
+        core.file.as_mut().unwrap().sync_all()?;
```
Comment (skyzh) [inline src/manifest.rs:291]: sync data should be enough.

File: src/manifest.rs [hunk 4]
```diff
@@ -0,0 +1,442 @@
...
+                    "manifest invalid, removing non-existing table {}",
+                    tc.id
+                )));
+            }
+        }
+    }
+    Ok(())
+}
+
+pub fn new_create_change(id: u64, level: usize, key_id: u64) -> ManifestChange {
```
Comment (skyzh) [inline src/manifest.rs:345]: I prefer to define this function on `ManifestChange`. e.g., ManifestChange::new_create, ManifestChange::new_delete

File: src/util.rs [hunk 5]
```diff
@@ -71,3 +71,14 @@ pub fn same_key(a: &[u8], b: &[u8]) -> bool {
     }
     return user_key(a) == user_key(b);
 }
+
+#[macro_export]
+macro_rules! write_with_length_check {
+    ($FILE:expr, $BUF:expr) => {
+        let buf_len = $BUF.len();
+        let written = $FILE.write(&$BUF)?;
```
Comment (skyzh) [inline src/util.rs:79]: Use [write_all](https://doc.rust-lang.org/std/io/trait.Write.html#method.write_all) instead?

**Review by skyzh — COMMENTED** [review]
Rest LGTM

File: src/manifest.rs (hunk 4 above)
Comment (GanZiheng) [reply src/manifest.rs:361]: `ManifestChange` is defined in external crate by `prost`, maybe we could not implement method for it directly?

Comment (skyzh) [reply src/manifest.rs:361]: Get it, that's true 🤣

File: src/error.rs [hunk 6]
```diff
@@ -26,6 +26,8 @@ pub enum Error {
     #[error("Invalid filename")]
     InvalidFilename(String),
     #[error("Invalid prost data: {0}")]
```
Comment (Connor1996) [inline src/error.rs:28]: Prost encode error:

File: src/manifest.rs [hunk 7]
```diff
@@ -0,0 +1,460 @@
...
+pub const MANIFEST_FILENAME: &str = "MANIFEST";
+const MANIFEST_REWRITE_FILENAME: &str = "MANIFEST_REWRITE";
+const MANIFEST_DELETION_REWRITE_THRESHOLD: usize = 10000;
+const MANIFEST_DELETIONS_RATIO: usize = 10;
+
+const MAGIC_TEXT: &[u8] = b"Agat";
+const MAGIC_VERSION: u32 = 8;
+
+/// LevelManifest contains information about LSM tree levels
+/// in the MANIFEST file.
```
Comment (Connor1996) [inline src/manifest.rs:27]: no need on a new line

File: src/manifest.rs [hunk 8]
```diff
@@ -0,0 +1,460 @@
...
+#[derive(Clone, Debug)]
+pub struct TableManifest {
+    pub level: u8,
+    pub key_id: u64,
+    // TODO: compression
+}
+
+/// Manifest represents the contents of the MANIFEST file.
+#[derive(Clone, Debug)]
+pub struct Manifest {
```
Comment (Connor1996) [inline src/manifest.rs:44]: where it is used

File: src/manifest.rs [hunk 9]
```diff
@@ -0,0 +1,460 @@
...
+// ManifestFile holds the file pointer (and other info) about the MANIFEST file, which is a log
+// file we append to.
+pub struct ManifestFile {
+    directory: PathBuf,
+    deletions_rewrite_threshold: usize,
+    core: Mutex<Core>,
+}
+
+impl Manifest {
+    pub fn new() -> Self {
```
Comment (Connor1996) [inline src/manifest.rs:77]: nothing special, can we derive `Default`?

File: src/manifest.rs [hunk 10]
```diff
@@ -0,0 +1,460 @@
...
+
+            apply_change_set(&mut build, &change_set)?;
+        }
+
+        Ok((build, offset))
+    }
+}
+
+impl ManifestFile {
+    pub fn open_or_create_manifest_file(opt: &AgateOptions) -> Result<Self> {
```
Comment (Connor1996) [inline src/manifest.rs:159]: please port the comment as well

File: src/manifest.rs (hunk 8 above)
Comment (GanZiheng) [reply src/manifest.rs:42]: It will be used in `db` and `levels` modules, which are not implemented yet.

**Review by Connor1996 — APPROVED** [review]
LGTM

**Pull request merged**